miner/worker.go

```diff
@@ -403,177 +243,91 @@ func (w *worker) newWorkLoop(recommit time.Duration) {
-			}
-
-			if w.resubmitHook != nil {
-				w.resubmitHook(minRecommit, recommit)
+			if cancel != nil {
+				cancel()
 			}
-
-		case <-w.exitCh:
-			return
-		}
-	}
-}
-
-// mainLoop is a standalone goroutine to regenerate the sealing task based on the received event.
-func (w *worker) mainLoop() {
-	defer w.txsSub.Unsubscribe()
-	defer w.chainHeadSub.Unsubscribe()
-	defer w.chainSideSub.Unsubscribe()
-
-	for {
-		select {
-		case req := <-w.newWorkCh:
 			if h, ok := w.engine.(consensus.Handler); ok {
 				h.NewWork()
 			}
-			w.commitNewWork(req.interrupt, req.noempty, req.timestamp)
-
-		case ev := <-w.chainSideCh:
-			// TOOO(nategraf): Remove this subcription, as there is no work to be done here.
-			log.Debug("Message in chan chainSideCh", "hash", ev.Block.Hash(), "number", ev.Block.Number(), "root", ev.Block.Root())
-
-		case ev := <-w.txsCh:
-			// Apply transactions to the pending state if we're not mining.
-			//
-			// Note all transactions received may not be continuous with transactions
-			// already included in the current mining block. These transactions will
-			// be automatically eliminated.
-			if !w.isRunning() && w.current != nil {
-				// If block is already full, abort
-				if gp := w.current.gasPool; gp != nil && gp.Gas() < params.TxGas {
-					continue
-				}
-				w.mu.RLock()
-				txFeeRecipient := w.txFeeRecipient
-				if !w.chainConfig.IsDonut(w.current.header.Number) && w.txFeeRecipient != w.validator {
-					txFeeRecipient = w.validator
-					log.Warn("TxFeeRecipient and Validator flags set before split etherbase fork is active. Defaulting to the given validator address for the coinbase.")
-				}
-				w.mu.RUnlock()
-
-				txs := make(map[common.Address]types.Transactions)
-				for _, tx := range ev.Txs {
-					acc, _ := types.Sender(w.current.signer, tx)
-					txs[acc] = append(txs[acc], tx)
-				}
-
-				txset := types.NewTransactionsByPriceAndNonce(w.current.signer, txs, w.createTxCmp())
-				tcount := w.current.tcount
-				w.commitTransactions(txset, txFeeRecipient, nil)
-				// Only update the snapshot if any new transactons were added
-				// to the pending block
-				if tcount != w.current.tcount {
-					w.updateSnapshot()
-				}
-			}
-			atomic.AddInt32(&w.newTxs, int32(len(ev.Txs)))
+			ctx, cancel = context.WithCancel(context.Background())
+			wg.Add(1)
+			go func() {
+				w.constructAndSubmitNewBlock(ctx)
+				wg.Done()
+			}()
 
 		// System stopped
 		case <-w.exitCh:
-			return
-		case <-w.txsSub.Err():
+			if cancel != nil {
+				cancel()
+			}
 			return
 		case <-w.chainHeadSub.Err():
-			return
-		case <-w.chainSideSub.Err():
+			if cancel != nil {
+				cancel()
+			}
 			return
 		}
 	}
 }
 
-// taskLoop is a standalone goroutine to fetch sealing task from the generator and
-// push them to consensus engine.
-func (w *worker) taskLoop() {
-	var (
-		stopCh chan struct{}
-		prev   common.Hash
-	)
+// interrupt aborts the in-flight sealing task.
+func (w *worker) interruptSealingTask() {
```

Comment:
take a look at my other comments, but if we follow what i suggested, where basically we have a "mainLoop" only, that creates a go routine that runs the whole block creation process, we can greatly simplify this, by using contexts.

The main loop will create a new context for each block, and it will cancel it on stop()